README.md
# Ratistics - Ruby Statistics Gem [![Build Status](https://secure.travis-ci.org/jdantonio/ratistics.png)](http://travis-ci.org/jdantonio/ratistics?branch=master) [![Dependency Status](https://gemnasium.com/jdantonio/ratistics.png)](https://gemnasium.com/jdantonio/ratistics)

*Ratistics* is a purely functional library that provides basic statistical
computation functions to Ruby programmers. It is intended for small data
sets only. This gem was designed for simplicity. Only basic consideration
was given to performance.

*Ratistics* functions operate any any enumerable object and support block
syntax for accessing complex data. This makes it possible to perform
statistical computations on a wide range of collections, including
ActiveRecord record sets.

Ratistics is pronounced *ra-TIS-tics*. Just like "statistics" but with an 'R'

    "Statistics".gsub(/^St/i, 'R') #=> "Ratistics"

The project is hosted on the following sites:

* [RubyGems project page](https://rubygems.org/gems/ratistics)
* [Source code on GitHub](https://github.com/jdantonio/ratistics)
* [YARD documentation on RubyDoc.org](http://rubydoc.info/github/jdantonio/ratistics/)
* [Continuous integration on Travis-CI](https://travis-ci.org/jdantonio/ratistics)
* [Dependency tracking on Gemnasium](https://gemnasium.com/jdantonio/ratistics)
* [Follow me on Twitter](https://twitter.com/jerrydantonio)

See the [changelog](https://github.com/jdantonio/ratistics/blob/master/CHANGELOG.md)
for information on the latest updates.

# About

A [friend](https://github.com/joecannatti) of mine convinced me
to learn statistics but I'm too lazy to learn [R](http://www.r-project.org/).
I started my statistics journey by reading the excellent book
[*Think Stats*](http://greenteapress.com/thinkstats/) by Allen B.
Downey and didn't want to do the exercises in Python. I looked for a
Ruby statistics library but couldn't find one that I liked.
So I decided to write my own.

## Project Goals

As much as possible I plan to follow these guidelines as I develop this gem
(in no particular order):

* Keep all functions pure and idempotent
* Support as many common collection classes as possible, especially ActiveRecord
* Keep runtime dependencies to a minimum, hopefully zero
* Remain backward compatable to Ruby version 1.9.2
* Support common Ruby interpreters (MRI, RBX, and JRuby)
* Be simple, consistent, and easy to use

### Purely Functional

This entire library is written in a purely functional style. All
functions are stateless, referentially transparent, and side-effect
free. When possible they are also idempotent. Only the sort order
of result sets may vary. The automated tests are run against
frozen data to ensure immutability.

Ruby does not support functions as first-class objects the way
purely functional languages do. Ruby's blocks are a pretty
close approximation. All the functions in this library accept
an optional block parameter. This supports computation against
complex data types without excessive data copying. I've followed
the same block idiom as much as possible to make the library as
consistent.
  
### Pure Ruby

Ratistics is written in pure Ruby and has no runtime gem dependencies.
It should work with any Ruby interpreter compliant with MRI 1.6.7 or
newer, but only MRI 1.9.x is officially supported.The test suite is
regularly run against several versions of Ruby so you should have
good results with any of the following:

* ruby-2.0.0
* ruby-1.9.3
* ruby-1.9.2
* jruby-1.7.0 (1.9 mode)
* jruby-1.6.8 (1.9 mode)
* rbx (1.9 mode)

#### ActiveRecord

[ActiveRecord](http://guides.rubyonrails.org/active_record_querying.html) is the
most common ORM in use by Ruby programmers. ActiveRecord result set collections
support the most common idiomatic Ruby enumerable methods. Subsequently, it
should be eacy to perform most calculations within this library directly against
ActiveRecord results set, using block syntax for field access. ActiveRecord is
not a runtime dependency of this gem but the test suite explicitly tests
ActiveRecord compatibality. 

#### Hamster

The main drawback of side-effect free functions is that in
non-functional languages they can lead to excessive data copying.
Some of the functions in this library, specifically ones that must
sort the data, may suffer from this. For better performance I highly
recommend using the [Hamster](https://github.com/jdantonio/hamster)
library of "efficient, immutable, thread-safe collection classes."
Hamster implements the Ruby [Enumerable](http://ruby-doc.org/core-1.9.3/Enumerable.html)
interface so all the functions in this library support the appropriate
Hamster classes. Hamster is not a runtime dependency of this gem but
the test suite explicitly tests Hamster compatibality.

## Installation

Install from RubyGems:

    gem install ratistics

or add the following line to Gemfile:

    gem 'ratistics'

and run *bundle install* from your shell.

## Usage

Require Ratistics within your Ruby project:

    require 'ratistics'

then use it:

    sample = [2, 3, 4, 5, 6]
    
    mean = Ratistics.mean(sample) #=> 4.0

When working with sets of complex data use blocks to process the data without copying:

    require 'active_record'

    class Person < ActiveRecord::Base; end

    people = Person.all
    
    mean = Ratistics.mean(people){|person| person.age}

### Statistics Functions

* cumulative_distribution_function (alias: cdf, cumulative_distribution)
* cumulative_distribution_function_value (alias: cdf_value, cumulative_distribution_value)
* first_quartile (alias: lower_quartile)
* frequency
* linear_rank (alias: linear_interpolation_rank)
* mean (alias: average, avg)
* median
* midrange (alias: midextreme)
* mode
* nearest_rank
* normalize_probability (alias: normalize_pmf)
* percent_rank (alias: percentile_rank)
* percentile (alias: centile)
* probability (alias: pmf)
* probability_mean (alias: pmf_mean, :frequency_mean)
* probability_variance (alias: pmf_variance)
* range
* ranks
* sample_with_replacement (alias: resample_with_replacement, bootstrap)
* sample_without_replacement (alias: resample_without_replacement, jackknife)
* second_quartile
* standard_deviation (alias: std_dev, stddev)
* third_quartile (alias: upper_quartile)
* truncated_mean (alias: trimmed_mean)
* variance (alias: var)

### I can drive that loader

Loading data from CSV and fixed field-width data files is a very common activity
in statistical computation. The methods in the Load module facilitate these
data loads. The methods in the Load module provide a robust syntax for
defining the individual fields in each record and processing the individual
fields on load.

    definition = [
      [:place, :to_i],
      nil,
      :div,
      :guntime,
      :nettime,
      :pace,
      nil,
      [:age, :to_i],
      :gender,
      [:race_num, :to_i],
    ]

    sample = Ratistics::Load::Csv.file('examples/race.csv', def: definition)
    sample.count #=> 1633
    sample.first #=> :place=>1, :div=>"M2039", :guntime=>"30:43", ... }

By default the methods of the Load module return Ruby Arrays. If the Hamster
gem is installed a Hamster collection can be returned instead. To return a
Hamster collection set the *:hamster* option to *true* or to a symbol
specifying the type to return. The default Hamster return type is
Hamster::Vector.

    require 'hamster'

    sample = Ratistics::Load::Csv.file('examples/race.csv', def: definition, as: :hash)
    sample.class #=> Array
    sample.first.class #=> Hash

    sample = Ratistics::Load::Csv.file('examples/race.csv', def: definition, as: :catalog)
    sample.class #=> Array
    sample.first.class #=> Array

    sample = Ratistics::Load::Csv.file('examples/race.csv', def: definition, as: :dataframe)
    sample.class #=> Array
    sample.first.class #=> Array

    sample = Ratistics::Load::Csv.file('examples/race.csv', def: definition, hamster: true)
    sample.class #=> Hamster::Vector

    sample = Ratistics::Load::Csv.file('examples/race.csv', def: definition, hamster: :set)
    sample.class #=> Hamster::Set

Consult the API documentation for the Load module for more information.

*I've got a Class Two rating.*

### Sorting

Some statistical computations require sorted data. In these cases
this library assumes the data is unsorted and calls the *#sort*
method on the data set. If the data passed to the function is
already sorted then an unnecessary performace penalty will occur.
To mitigate this, every function that requires sorted data provides
an optional *sorted* parameter which defaults to false. When set
to *true* it indicates the data is already sorting and the sort
step is skipped.

A problem occurs when the data set does not support a natural sort
order. The Ruby idiom for this situation is to accept a block which
specifies the sort operation. Unfortunately, passing two blocks to
a function is cumbersome in Ruby so the functions that require sorting
must depend on the natural sort order only. For simplicity and
consistency, when a block is passed to a function that requires
sorted data it is assumed that natural sorting is impossible.
Subsequently the sort operation is skipped regardless of the value
of the *sorted* parameter.


### A Worked Example

The following code answers the big question from page 2 of *Think Stats*: Do first babies arrive late?

The first step in this solution is to load the NSFG test data from the 'examples'
directory. The Ratistic::Load module is used to load a subset of the fields from each
record. Each record represents a single pregnancy and has the following structure:

    {:caseid => '1',
     :nbrnaliv => '1',
     :babysex => '1',
     :birthwgt_lb => '8',
     :birthwgt_oz => '13',
     :prglength => '39',
     :outcome => '1',
     :birthord => '1',
     :agepreg => '3316',
     :finalwgt => '6448.271111704751'}

Once the data is loaded it can be easily processed:

    # load the data
    fields = [
      {:field => :caseid, :start => 1, :end => 12},
      {:field => :nbrnaliv, :start => 22, :end => 22},
      {:field => :babysex, :start => 56, :end => 56},
      {:field => :birthwgt_lb, :start => 57, :end => 58},
      {:field => :birthwgt_oz, :start => 59, :end => 60},
      {:field => :prglength, :start => 275, :end => 276},
      {:field => :outcome, :start => 277, :end => 277},
      {:field => :birthord, :start => 278, :end => 279},
      {:field => :agepreg, :start => 284, :end => 287},
      {:field => :finalwgt, :start => 423, :end => 440},
    ]
    sample = Ratistics::Load.dat_gz_file('data/2002FemPreg.dat.gz', fields)
    sample.count #=> 13593 

    # filter for first-borns
    first = sample.select{|item| item[:birthord].to_i == 1}
    first.count #=> 4413

    # filter for non-first-borns
    not_first = sample.select{|item| item[:birthord].to_i > 1}
    not_first.count #=> 4735

    # calculate mean pregnancy lengths
    Ratistics.mean(sample){|item| item[:prglength]} #=> 29.531229309203265 
    Ratistics.mean(first){|item| item[:prglength]} #=> 38.60095173351461 
    Ratistics.mean(not_first){|item| item[:prglength]} #=> 38.52291446673706

    # calculate the variance of pregnancy lengths
    Ratistics.variance(sample){|item| item[:prglength]} #=> 190.49562224367648 
    Ratistics.variance(first){|item| item[:prglength]} #=> 7.792947202066306 
    Ratistics.variance(not_first){|item| item[:prglength]} #=> 6.84123839078341

    # calculate the standard deviation of pregnancy lengths
    Ratistics.standard_deviation(sample){|item| item[:prglength]} #=> 13.8020151515522 
    Ratistics.standard_deviation(first){|item| item[:prglength]} #=> 2.7915850698243654 
    Ratistics.standard_deviation(not_first){|item| item[:prglength]} #=> 2.6155761106844913

    # calculate the frequency of pregnancy lengths
    sample_freq = Ratistics.frequency(sample){|item| item[:prglength]} #=> {"39"=>4744, "38"=>609, ...}
    first_freq = Ratistics.frequency(first){|item| item[:prglength]} #=> {"39"=>2114, "38"=>272, ...}
    not_first_freq = Ratistics.frequency(not_first){|item| item[:prglength]} #=> {"39"=>2579, "40"=>580, ...}

Once you have the frequency data you can use any charting/graphing library to
create a histogram to compare birth rates. The file
[histogram.rb](https://github.com/jdantonio/ratistics/blob/master/examples/histogram.rb)
shows how to use [Gruff](https://github.com/topfunky/gruff) to create this:

![histogram](https://raw.github.com/jdantonio/ratistics/master/examples/histogram.png)

## Test Data

The test data shipped with this gem is freely available from the Centers for Disease Control
and Prevention [National Survey of Family Growth](http://www.cdc.gov/nchs/nsfg.htm) and from
the [27th Anniversary Edition James Joyce Ramble 10K](27th Anniversary Edition James Joyce Ramble 10K).

## Copyright

*Ratistics* is Copyright &copy; 2013 [Jerry D'Antonio](https://twitter.com/jerrydantonio).
It is free software and may be redistributed under the terms specified in
the LICENSE file.

## License

Released under the MIT license.

http://www.opensource.org/licenses/mit-license.php  

> Permission is hereby granted, free of charge, to any person obtaining a copy  
> of this software and associated documentation files (the "Software"), to deal  
> in the Software without restriction, including without limitation the rights  
> to use, copy, modify, merge, publish, distribute, sublicense, and/or sell  
> copies of the Software, and to permit persons to whom the Software is  
> furnished to do so, subject to the following conditions:  
> 
> The above copyright notice and this permission notice shall be included in  
> all copies or substantial portions of the Software.  
> 
> THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR  
> IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,  
> FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE  
> AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER  
> LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,  
> OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN  
> THE SOFTWARE.  


CHANGELOG.md
# Ratistics Changelog

### 0.3.0 *(in progress)*

Release Date: TBD

* Dropped support for Ruby < 1.9.2
* Completely refactored data loading helper modules
  * Significantly increased performance when loading CSV files
* Removed classes Aggregates, Frequencies, and Percentiles
* Removed monkey patching code

### 0.2.3

Release Date: March 22, 2013

* Added functions
  * Probability.cumulative_distribution_function (alias: cdf, cumulative_distribution)
  * Probability.cumulative_distribution_function_value (alias: cdf_value, cumulative_distribution_value)
  * Probability.sample_with_replacement (alias: resample_with_replacement, bootstrap)
  * Probability.sample_without_replacement (alias: resample_without_replacement, jackknife)
  * Frequencies.cumulative_distribution_function (alias: cdf, cumulative_distribution)
  * Frequencies.cumulative_distribution_function_value (alias: cdf_value, cumulative_distribution_value)
  * Array.sample_with_replacement (alias: resample_with_replacement, bootstrap)
  * Array.sample_without_replacement (alias: resample_without_replacement, jackknife)
  * Collection.collect
  * Collection.catalog (alias: catalogue)
  * Collection.catalog_hash(alias: catalogue_hash)
  * Collection.hash_catalog(alias: hash_catalogue)
  * Collection bisection
    * bisect_left, bisect_right (alias: bisect)
    * insort_left, insort_right (alias: insort)
    * insort_left!, insort_right! (alias: insort!)
* Added Inflect module
* Added NilSampleError
* Added Catalog (alias: Catalogue) class
* Greater consistency of return values and processing options
* Added more input and output options:
  * :as => :hash/:map/:array/:catalog/:catalogue to probability functions
  * :as => :hash/:map/:array/:catalog/:catalogue to percentile functions
  * :from => :sample/frequency/probability to frequency and probabilty functions
  * :inc/:increment/:incremental => true/false to #probability function
* Changed return values for Rank.ranks
  * returns nil when the sample is nil or empty
  * returns :hash by default
  * only returns an array when :as => :array
* Added Infinity and NaN constants

### 0.2.2

Release Date: February 14, 2012

* Added classes
  * Aggregares
  * Frequencies
  * Probabilities
* Added percentile functions
  * percentile
  * first_quartile (alias: lower_quartile)
  * second_quartile
  * thirs_quartile (alias: upper_quartile)
* Added math functions
  * summation (alias: sum)
* Added sorts and searches
  * insertion_sort!
  * linear_search

### 0.2.1

Release Date: February 4, 2013

Removed runtime dependencies that were accidentally added in 2.0.0.
They were intended to be development dependencies only.

### 0.2.0

Release Date: February 4, 2013

This release greatly enhanced the tests by adding support for ActiveRecord.
It also greatly enhanced cross-Ruby compatibility by fixing code that
was broken under 1.8.7 and by verifying compatibility with Rubinius.
Finally, several arithmetic and statitical computations.

* Added *Rank* module with functions related to percentiles
  * ranks (alias: percentiles, centiles)
  * linear_rank (aias: percentile, centile)
  * nearest_rank (with :ordinal, :nist_primary, and :nist_alternate formulas)
  * percent_rank
* Added math functions with block support
  * min
  * max
  * minmax
  * relative_risk
* Added collections functions with block support
  * ascending?
  * descending?
  * binary_search (alias: bsearch, half_interval_search)
* Added *#midrange* function to the *Average* module
* Added tests to confirm ActiveRecord compatability
* Refactored CSV loading for better support across Ruby versions
* Updated all functions with a *sorted* parameter to use a *:sorted* option instead
* Added an options hash to all stats functions
* Added more monkey-patch methods
* Added a *#slice* function to support more collection classes
* Support Hamster classes as return types when loading data from files
* Began testing under the Rubinius interpreter
* Bagan documenting answers to exercises to *Think Stats* in the *examples* directory
* Added more sample data and sample code to the *examples* directory
* Updated documentation


examples/survey.rb
require 'zlib'
require 'hamster/vector'
require 'csv'


# sample = [5, 1, 9, 3, 14, 9, 7]
#
#  1 - 7/0/1/7.142857142857142
#  2 - 7/1/0/14.285714285714285
#  3 - 7/1/1/21.428571428571427
#  4 - 7/2/0/28.57142857142857
#  5 - 7/2/1/35.714285714285715
#  6 - 7/3/0/42.857142857142854
#  7 - 7/3/1/50
#  8 - 7/4/0/57.14285714285714
#  9 - 7/4/2/71.42857142857143
# 10 - 7/6/0/85.71428571428571
# 11 - 7/6/0/85.71428571428571
# 12 - 7/6/0/85.71428571428571
# 13 - 7/6/0/85.71428571428571
# 14 - 7/6/1/92.85714285714286
#
# sample = [40, 15, 35, 20, 40, 50]
#
#  5 - 6/0/0/0
# 10 - 6/0/0/0
# 15 - 6/0/1/8.333333333333332
# 20 - 6/1/1/25
# 25 - 6/2/0/33.33333333333333
# 30 - 6/2/0/33.33333333333333
# 35 - 6/2/1/41.66666666666667
# 40 - 6/3/2/66.66666666666666
# 45 - 6/5/0/83.33333333333334
# 50 - 6/5/1/91.66666666666666
# 55 - 6/6/0/100  
#
# sample = [1,1,1,1,1,2,2,2,2,2,2,2,2,3,3,3,3,3,4,4,5,6]
#
# 1 - 22/0/5/11.363636363636363
# 2 - 22/5/8/40.909090909090914
# 3 - 22/13/5/70.45454545454545
# 4 - 22/18/2/86.36363636363636
# 5 - 22/20/1/93.18181818181817
# 6 - 22/21/1/97.72727272727273
#
# Formula:
# PR% = L + ( 0.5 x S ) / N 
#
# Where,
# L = Number of below rank, 
# S = Number of same rank,
# N = Total numbers.

module Survey
  extend self

  FEMRESP_RAW = File.expand_path(File.join(File.dirname(__FILE__), '2002FemResp.dat.gz'))
  FEMRESP_CSV = File.expand_path(File.join(File.dirname(__FILE__), '2002FemResp.csv'))

  FEMPREG_RAW = File.expand_path(File.join(File.dirname(__FILE__), '2002FemPreg.dat.gz'))
  FEMPREG_CSV = File.expand_path(File.join(File.dirname(__FILE__), '2002FemPreg.csv'))

  RACE_RAW = File.expand_path(File.join(File.dirname(__FILE__), 'race.dat.gz'))
  RACE_CSV = File.expand_path(File.join(File.dirname(__FILE__), 'race.csv'))

  RESPONDENT_FIELDS = [
    [:caseid, 1, 12, :to_i].freeze,
  ].freeze

  PREGNANCY_FIELDS = [
    [:caseid, 1, 12, :to_i].freeze,
    [:nbrnaliv, 22, 22, :to_i].freeze,
    [:babysex, 56, 56, :to_i].freeze,
    [:birthwgt_lb, 57, 58, :to_i].freeze,
    [:birthwgt_oz, 59, 60, :to_i].freeze,
    [:prglength, 275, 276, :to_i].freeze,
    [:outcome, 277, 277, :to_i].freeze,
    [:birthord, 278, 279, :to_i].freeze,
    [:agepreg, 284, 287, :to_i].freeze,
    [:finalwgt, 423, 440, :to_f].freeze,
  ].freeze

  RACE_FIELDS = [
    [:place, 1, 6, :to_i].freeze,
    [:div_tot, 7, 15, :strip].freeze,
    [:string, 16, 21, :strip].freeze,
    [:guntime, 22, 29, :strip].freeze,
    [:nettime, 30, 38, :strip].freeze,
    [:pace, 39, 44, :strip].freeze,
    [:name, 45, 67, :strip].freeze,
    [:age, 68, 70, :to_i].freeze,
    [:gender, 71, 72, :strip].freeze,
    [:race_num, 73, 78, :to_i].freeze,
    [:location, 79, 101, :strip].freeze,
  ].freeze

  def get_respondent_data
    @@respondents ||= load_file(Survey::FEMRESP_RAW, Survey::RESPONDENT_FIELDS)
  end

  def get_pregnancy_data
    @@pregnancies ||= load_file(Survey::FEMPREG_RAW, Survey::PREGNANCY_FIELDS)
  end

  def get_race_data
    @@race ||= load_file(Survey::RACE_RAW, Survey::RACE_FIELDS)
  end

  def load_respondent_csv
    @@respondents_csv ||= load_csv(Survey::FEMRESP_CSV, Survey::RESPONDENT_FIELDS)
  end

  def load_pregnancy_csv
    @@pregnancies_csv ||= load_csv(Survey::FEMPREG_CSV, Survey::PREGNANCY_FIELDS)
  end

  def load_race_csv
    @@race_csv ||= load_csv(Survey::RACE_CSV, Survey::RACE_FIELDS)
  end

  def get_counts
    respondents = get_respondent_data
    puts "Number of respondents: #{respondents.count}"

    pregnancies = get_pregnancy_data
    puts "Number of pregnancies: #{pregnancies.count}"

    racers = get_race_data
    puts "Number of racers: #{racers.count}"
  end

  def load_file(path, definition)
    data = Hamster.vector

    Zlib::GzipReader.open(path) do |gz|
      gz.each_line do |line|
        record = {}
        definition.each do |field|
          record[field[0]] = line.slice(field[1]-1, field[2]-field[1]+1).send(field[3])
        end
        data = data.cons(record.freeze)
      end
    end

    return data
  end

  def load_csv(path, definition)
    data = Hamster.vector

    CSV.foreach(path) do |row|
      record = {}
      definition.each_index do |index|
        record[definition[index][0]] = row[index].send(definition[index][3])
      end
      data = data.cons(record.freeze)
    end

    return data
  end
end
